src/checkqueue.h

```diff
@@ -30,103 +30,112 @@ template <typename T>
+                    fMasterPresent = false;
+                    // There's no harm to the master holding the lock
+                    // at this point because all the jobs are taken.
+                    // so busy spin until no one else is awake
+                    while (nAwake) {}
+                    bool fRet = fAllOk;
+                    // reset the status for new work later
+                    fAllOk = 1;
+                    // return the current status
+                    return fRet;
+                } else  if (!fMasterPresent) { // Read once outside the lock and once inside
+                    --nAwake; 
+                    // Unfortunately we need this lock for this to be safe
+                    // We hold it for the min time possible
+                    {
+                        boost::unique_lock<boost::mutex> lock(mutex);
+                        condWorker.wait(lock, [&]{ return fMasterPresent.load();});
                     }
-                    nIdle++;
-                    cond.wait(lock); // wait
-                    nIdle--;
+                    ++nAwake;
                 }
-                // Decide how many work units to process now.
-                // * Do not try to do everything at once, but aim for increasingly smaller batches so
-                //   all workers finish approximately simultaneously.
-                // * Try to account for idle jobs which will instantly start helping.
-                // * Don't do batches smaller than 1 (duh), or larger than nBatchSize.
-                nNow = std::max(1U, std::min(nBatchSize, (unsigned int)(check_mem_top - check_mem_bottom) / (nTotal + nIdle + 1)));
-                checks_iterator = check_mem_bottom;
-                std::advance(check_mem_bottom, nNow);
-                // Check whether we need to do work at all
+            } else {
+                // We compute using bottom_cache (not bottom_cache + 1 as above) because it is 0-indexed
+                T * pT = check_mem + bottom_cache;
+                // Check whether we need to do work at all (can be read outside
+                // of lock because it's fine if a worker executes checks
+                // anyways)
                 fOk = fAllOk;
-            }
-            // execute work
-            for (unsigned int i = 0; i < nNow && fOk; i++) {
-                fOk = (*checks_iterator)();
+                // execute work
+                fOk &= fOk && (*pT)();
+                // We swap in a default constructed value onto pT before freeing
+                // so that we don't accidentally double free when check_mem is
+                // freed. We don't strictly need to free here, but it's good
+                // practice in case T uses a lot of memory.
                 auto t = T();
-                checks_iterator->swap(t);
-                std::advance(checks_iterator, 1);
+                pT->swap(t);
+                // Can't reveal result until after swapped, otherwise
```

Comment:
Yes; you're correct. Earlier version was like this.

Also termination does still occur "early", all the actual checks are skipped (they are just stilled dequeued). I suppose I could make it abort (may actually be nice to check if it aborted before calling checkinputs... but maybe that's best for another PR).

I'll refactor to something which converges more quickly on abort (e.g., setting fMasterPresent && !fAllOk)